### ISSUE TITLE
Support hmac-secret FIDO2 extension for calculating hash

### ISSUE THREAD

**Issue by antelle** [keeweb/keeweb#1681]
KeeWeb connects to YubiKeys using their proprietary HMAC-SHA1 Challenge-Response API, which is less than ideal. FIDO2 standard now includes [`hmac-secret`](https://fidoalliance.org/specs/fido-v2.0-rd-20180702/fido-client-to-authenticator-protocol-v2.0-rd-20180702.html#sctn-hmac-secret-extension) extension, which provides similar functionality, but implemented in a standard way.

This should give us support for other tokens, for example, Trezor One, without using their API.  
Additionally, if `hmac-secret` is ever supported in browsers (for example, [this PR](https://bugzilla.mozilla.org/show_bug.cgi?id=1551594) adds support for it, but it's limited to Windows Hello), we'll be able to use it on web as well.

Related issue in KeePassXC: https://github.com/keepassxreboot/keepassxc/issues/3560
`libfido2`: https://github.com/Yubico/libfido2
`hmac-secret` usage: https://github.com/trezor/trezor-firmware/blob/master/tests/fido_tests/libfido2/hmac-secret.sh

**Comment by rugk** [comment]
Regarding [the current campaign for SoloKey v2](https://www.kickstarter.com/projects/conorpatrick/solo-v2-safety-net-against-phishing) there has been [some comment by the creators](https://github.com/solokeys/kickstarter2021/discussions/23#discussioncomment-312837) that supports this implementation, too:
> We still think the way to do this is to use the `hmac-secret` extension that is part of the actual standard (it is in essence HMAC-SHA256 challenge-response). In that sense, the answer is, yes, at launch (and Solo V1 has it too, just like every fully certified FIDO2 dongle).
> If the question is whether we will implement Yubico's proprietary app (in essence, HMAC-SHA1 challenge-response), the answer is no, we do not plan to do so. This sentiment is shared by other vendors such as Trezor - if OSS such as KeePassXC and LUKS moves on to an actual standard instead of one company's private sauce, everyone wins.

**Comment by gynet** [comment]
Very cool. Once it got implemented, is keeweb going to deprecate the existing 'Challenge-Response'?

**Comment by antelle** [comment]
Depends on its compatibility and ecosystem progress. My gut feeling says it may be not interchangeable with the existing mechanism, as in, it will generate different codes, and most apps will probably support both methods. But nobody knows now...

**Comment by leandroferreira** [comment]
@antelle is this on keeweb's roadmap? Is there appetite for someone from the community to give it a go at implementing this solution?

**Comment by antelle** [comment]
First step is to understand how it actually works. Help here would be very welcome and if the understanding is present, we can progress from here.

**Comment by antelle** [comment]
Indeed, using libfido2

**Comment by antelle** [comment]
I did some investigation, and the script linked in the issue ([hmac-secret.sh](https://github.com/trezor/trezor-firmware/blob/master/tests/fido_tests/libfido2/hmac-secret.sh)) works well with a YubiKey with a [small correction](https://gist.github.com/antelle/20981b15dcac5d131d44f25108ea10b9) for a combination macOS and YubiKey. It generates equal results for the same input.

Now, some thoughts:

1. It's not possible to change or manually set the **seed** (unlike the Yubico's implementation), which means, you won't be able to program another YubiKey with the same secret. For now, I'm not sure what to do with this and it's a blocker for adoption.
2. (1.) means it's obviously **not compatible with the existing implementation**. The usage of `hmac-secret` FIDO2 extension also not implemented in any of forks, so we don't have anything to compare.
3. Discussion in the linked issue make me think that libfido2 **doesn't work on Windows**, however there seems to be a workaround (use Windows API), if the workaround works, it should be fine, just a bit more code.
4. I'm also not sure if this will work on **iOS**, but there's a possibility that it will. I don't have a possibility to check it though, as far as I know you need to have MFi for testing this. Because of (1.), there won't be an option to unlock the database using just the seed, without YubiKey, which means Safari integration won't work.

So, now the biggest question is, how to provide **support for multiple keys**? Without backup, using a YubiKey with kdbx doesn't make a lot of sense. I can imagine something like there's another key stored in the db that is used for backups. When you save the db, it makes a backup that can be opened with that separate key. But it still doesn't sound right and probably it won't be implemented like this across forks.

**Comment by leandroferreira** [comment]
Thank you for looking into this, it certainly sounds promising!
I agree with your premise, and I'm not completely sure what would be the steps for validating solution 2 (honestly, some of it flew over my head). Sorry if I'm not being much help.

**Comment by antelle** [comment]
I spent some time thinking about it today. It's going to be a long read, sorry, I tried to put as much detail as I could.

First, let me give some context. As it's mentioned above, [FIDO2 `hmac-secret`](https://fidoalliance.org/specs/fido-v2.1-rd-20210309/fido-client-to-authenticator-protocol-v2.1-rd-20210309.html#sctn-hmac-secret-extension) extension provides support for hashing arbitrary data using a secret attached to a credential stored on the authenticator. It has two inputs:

1. credential id generated by the authenticator, obtained during credential registration (512 bit)
2. challenge, client-defined data (512 bit)

Some notes to clarify the implementation:

- The secret used for signing data is stored on the authenticator and is never exposed. It's not possible to provide your own secret when making a new credential. This means, you can't make a backup of your YubiKey that would produce the same hash based on the same input.
- Credential id must be generated by the authenticator, it's not possible to make your own one. Even if it's non-resident (see below), the authenticator must not accept credentials produced not by this authenticator.
- FIDO2.1 defines discoverable (aka "resident") credentials. To retrieve such credential, most of authenticators will present the PIN question.
    - benefit: we don't have to store credential id (however we still need to identify the authenticator, so there's not much benefit here as it's also noted in the linked issue in KeePassXC and we can just use regular, not persistent credentials)
    - downside: space for these credentials is very limited, it's usually around 5..20 credentials per authenticator in total, if it's implemented at all
- `hmac-secret` is a FIDO2 assertion, most of authenticators check user presence, for example, on a YubiKey you need to press the button.

In the KDBX format, such kind of hashing is used to calculate the master hash based on salt stored in the public file header. Let's assume, we want to protect our database with a password and a YubiKey. To open the database, we need both things. In this case the master key is calculated as follows (this explanation is simplified to illustrate the idea):

```
MasterKey = KDF( PasswordHash ⨁ HMAC(Challenge) )
```

where:
- `KDF` is a so-called key derivation function, one-way resource-hard transformation, such as Argon2
- `PasswordHash` is calculated based on entered password
- `HMAC` is the transformation performed by the authenticator
- `Challenge` is something stored in the public file header
- `MasterKey` is the encryption key used to decrypt data stored in the database

Our goal is to have severeal authenticators each of them would be able to open the database. As it's not possible to program your own secret, these authenticators can't have equal `HMAC` transformation and instead they provide `HMAC_1`, `HMAC_2`, and so on.

We want to find a way to make such challenge that:
```
MasterKey = KDF( PasswordHash ⨁ HMAC_1(Challenge) )
MasterKey = KDF( PasswordHash ⨁ HMAC_2(Challenge) )
...
MasterKey = KDF( PasswordHash ⨁ HMAC_N(Challenge) )
```

To achieve this, we add an additional step to `HMAC` function that depends on used authenticator and an additional value stored in the header (let's call it `HMAC*`):

```
HMAC*: (data, x) => {
    TransformResponse(
        HMAC_X(
            MakeChallenge(data, x)
        ),
        data,
        x
    )
}

MakeChallenge: (data, x) => {
    item.challenge from data
        where AuthenticatorXHasIdentity(item.identity)
}

TransformResponse: (response, data, x) => {
    response ⨁ (item.transform from data
                    where AuthenticatorXHasIdentity(item.identity))
}
```

where
- `x` is authenticator id
- `data` is a data structure stored in the public header, see below

Addition to the header presented in human-readable form (would be stored in the [header extensions](https://keepass.info/help/kb/kdbx_4.html#exthdr) part as binary, json is only for explanation):
```json
{
    "auth": [
        {
            "identity": "credential identification, for example, credential_id",
            "challenge": "challenge used to compute HMAC on the authenticator",
            "transform": "piece of data used for HMAC transformation"
        }
    ]
}
```

When saving the file, we need to generate a new challenge-response on this authenticator. It can be achieved by updating `transform` value for other authenticators used to decrypt this file, so that HMAC_X calculation will produce the same result:

```
ReplaceHmac: (data, x) => {
    // this can be cached to avoid computation
    old_hmac := HMAC*(data, x)

    data[x].challenge := random()
    data[x].transform := random()

    new_hmac := HMAC*(data, x)

    authenticator_i_response := data[i].transform ⨁ old_hmac
    data[i].transform := new_hmac ⨁ authenticator_i_response
        for i in data where i ≠ x
}
```

Additionally, there can be a section inside the file where you can assign names to authenticators. This will allow users to detach authenticators from files. This section will contain the mapping of `data.identity` to a human-readable representation.

Here's an image representation of the proposed change:

![hmac-secret in kdbx](https://user-images.githubusercontent.com/633557/113034278-35df2000-9192-11eb-9a73-b1770421f6bc.png)


Benefits of this approach:
1. Users can link any number of authenticators to one database;
2. YubiKeys don't have to be re-programmed manually, so the probability of secret leakage is less;
3. It can be extensible for other authentication methods, for example, WebAuthn on web.

Concerns:
1. Storing identities in the public header of the database exposes used authenticators, if attackers get access to the kdbx file, they will know that you use an authenticator to open it. I'm not sure if this is a concern, I need advice here.
2. It would be good to have it defined centralized in the format. When it's spread across different implementations, it's error-prone.

How to proceed from here? For now I need some feedback and critique, maybe the whole idea of having publicly available `identity` and `challenge` is absurd. It makes sense to get an approval from different popular KeePass implementations before we write any code.

**Comment by rugk** [comment]
In any case, it should be compatible with KeePassXC etc.

**Comment by antelle** [comment]
@rugk it won't be compatible with the current implementation. I assume KeePassXC, as well as the 2FA KeePass plugin, will come up with a new variant. I hope it will be accepted in the kdbx format, but I don't know, I can ask Dominik.

**Comment by antelle** [comment]
I've moved it to Future not because I find it unnecessary, but it's not very much actionable in the current state. Now we need to come up with a solution that would work and agree on this, it won't happen fast.

**Comment by gynet** [comment]
> The secret used for signing data is stored on the authenticator and is never exposed. It's not possible to provide your own secret when making a new credential. 

This makes me feel sad about the standard

**Comment by antelle** [comment]
Well, it is good! The standard is here to protect us and make our world better. If secrets can leak, they will leak. If secrets can’t leak, developers will adapt and eventually make the software accept several authenticators, it will be better for users, right?
A real issue here is the kdbx format. If it was my software with mobile apps, etc... multiple authenticator enrollment would be a matter of couple of meetings. Now... 🤷‍♀️  who knows when other developers want to support it.

**Comment by gynet** [comment]
Yeah, it's more secure, but to be honest, I like how the current Yubikey challenge-response or GPG works: I can generate or backup my secret or key in two USB drives using my secure laptop, and lock up them in a safebox, and it is more flexible, but sometimes I am cynical, and I do see the business reasons might behind "standard makers" :P   

Anyway, it seems like if keeweb applies the FIDO2 standard, users have to purchase at least two physical authenticators in order to prevent lost/theft scenarios, I think it's might be good for online services, because usually, they have other alternative ways to finish authentication or reset the key, but for offline auth or decryption, have a backup physical authenticator might be the only way to recover a database.

**Comment by antelle** [comment]
Yes, that's the point, that you have a backup authenticator. It makes sense to have it anyway if you use WebAuthn.

But this can be made a soft requirement. If multiple authenticators are supported in KDBX, one of them may be generation based on a backup code, or stored on a server and given via WebAuthn, and so on.

**Comment by LunarEclipse363** [comment]
For web you want to use the [WebAuthn PRF extension](https://github.com/w3c/webauthn/wiki/Explainer:-PRF-extension) ([MDN page](https://developer.mozilla.org/en-US/docs/Web/API/Web_Authentication_API/WebAuthn_extensions#prf)) which uses hmac-secret under the hood.
